Bugfix/lattice 2318 linking indexing distributed queue destroyed

you-leee opened the pull request.

Commit:
LATTICE-2318 use poll instead of take, make indexing scheduled, query improvements

Commit:
LATTICE-2318 tombstoned entity won't be linked

geekbeast commented:
I would have to take a closer look at the code, but the immediate thing that caught my attention is that a queue is being used a long with lock/unlock mechanism. 

One of the points of using a distributed queue is so that work can be pulled from it without having to worry about locks. 

The other thing that caught my attention is the sequences are generated multiple times by polling the queue? That seems strange, but it might be just the snippet I can see in the PR.

you-leee commented:
Yeah, I could remove the linking id locking part.

So, the original problem was, that the distributed queues got destroyed (https://docs.hazelcast.org/docs/3.10/javadoc/com/hazelcast/spi/exception/DistributedObjectDestroyedException.html) right at the beginning of indexer startup, because there were no objects in the queue and take() threw an Exception. Then I changed it to poll(), but the problem with that was, that the queues were never read, only written (with the same linking ids multiple times) periodically. So that's when I changed it to create the sequence also periodically and locally it works fine.
I'm still not sure, why creating the stream out of a chunked sequence is throwing this exception, but directly creating it from the queue not.

I'm up to any suggestions at this point.

you-leee commented:
For easier review, these are the 2 queries used to collect linking ids (+ entity data keys + last write) for indexing and un-indexing:
```
WITH valid_linking_entities AS
  (SELECT linking_id,
          id,
          entity_set_id,
          last_write,
          last_index,
          last_link,
          last_link_index
   FROM ids
   WHERE linking_id IS NOT NULL
     AND VERSION > 0)
SELECT linking_id,
       array_agg(ARRAY[entity_set_id, id]) AS entity_data_key,
       max(last_write) AS last_write
FROM valid_linking_entities
WHERE last_index >= last_write
  AND last_link >= last_write
  AND last_link_index < last_write
GROUP BY linking_id
UNION ALL
SELECT linking_id,
       array_agg(ARRAY[entity_set_id, id]) AS entity_data_key,
       max(last_write) AS last_write
FROM ids
WHERE linking_id IS NOT NULL
  AND VERSION <= 0
  AND last_index >= last_write
  AND last_link_index < last_write
  AND linking_id IN
    (SELECT linking_id
     FROM valid_linking_entities)
  AND entity_set_id IN
    (SELECT entity_set_id
     FROM valid_linking_entities)
GROUP BY linking_id
LIMIT 3000;


SELECT linking_id,
       array_agg(ARRAY[entity_set_id, id]) AS entity_data_key,
       max(last_write) AS last_write
FROM ids
WHERE linking_id NOT IN
    (SELECT linking_id
     FROM ids
     WHERE linking_id IS NOT NULL
       AND VERSION > 0)
  AND linking_id IS NOT NULL
  AND last_index >= last_write
  AND last_link_index < last_write
GROUP BY linking_id
LIMIT 3000;
```

geekbeast commented:
Yikes! For (non)starters, citus doesn't really have cursors and will load every linking id into memory for that subquery on every shard, move it to the coordinator node then push it down to all other nodes. 

We need one job limit based on id or linking_id that only queues for processing N linking ids at a the time.

you-leee commented:
Update: I've let the background linking service running for a while and it throws the same error.. but it still manages to put/take into/from queue and link after I create new entities to be linked.

Commit:
Merge branch 'develop' into bugfix/LATTICE-2318-linking-indexing-distributed-queue-destroyed

* develop:
  use Organizations IMap instead of OrganizationTitles IMap

Commit:
maybe this way it works

Commit:
compile

Commit:
LATTICE-2318 go back to scheduled tasks for linking indexing, limit queries, add more logging

Commit:
func order

katherinebernstein commented on src/main/kotlin/com/openlattice/indexing/BackgroundLinkingIndexingService.kt:
```diff
@@ -103,54 +105,117 @@ class BackgroundLinkingIndexingService(
-                    }
+    @Scheduled(fixedRate = LINKING_INDEX_RATE)
+    private fun linkingIndexing() {
+        if (isLinkingIndexingEnabled()) {
```
I'd prefer this as 
```
if (isLinkingIndexingEnabled()) {
  return
}

executor.submit {
  ...
}
```

just to reduce the amount of nesting here, since it gets pretty deep

katherinebernstein commented on src/main/kotlin/com/openlattice/indexing/BackgroundLinkingIndexingService.kt:
```diff
@@ -103,54 +105,117 @@ class BackgroundLinkingIndexingService(
-                    }
+    @Scheduled(fixedRate = LINKING_INDEX_RATE)
+    private fun linkingUnIndexing() {
+        if (isLinkingIndexingEnabled()) {
```
Same comment as above

katherinebernstein replied:
(and same comment below)

katherinebernstein replied:
This also looks nearly identical to the function above -- it seems preferable to combine these into one

katherinebernstein commented on src/main/kotlin/com/openlattice/indexing/BackgroundLinkingIndexingService.kt:
```diff
@@ -103,54 +105,117 @@ class BackgroundLinkingIndexingService(
+    @Suppress("UNUSED")
+    @Scheduled(fixedRate = LINKING_INDEX_RATE)
+    private fun updateUnIndexCandidates() {
+        if (isLinkingIndexingEnabled()) {
```
seems like this can also be combined with `updateIndexCandidates` (and same comment about unnesting)

Commit:
LATTICE-2318 early return, fix expiration check, add listener to eviction

dbt commented on src/main/kotlin/com/openlattice/indexing/BackgroundLinkingIndexingService.kt:
```diff
@@ -284,23 +325,58 @@ class BackgroundLinkingIndexingService(
+
+    /**
+     * Checks if linking id is already "locked" in [linkingIndexingLocks] map.
+     * If it's not, it returns true, if it is, the function refreshes the expiration of the entry and returns false.
```
I'm not sure, having read this comment, what a true or false value means for the caller. I realize it's private impl and I'm still learning the codebase, but something like "returns true if the object is available to be processed, and false if it's already locked by another worker" would be helpful.

you-leee replied:
Made some changes for the doc here, hope it's more understandable this way. The basic idea would be: a linking entity will be locked if it's up for processing for either indexing or un-indexing (should be mutually exclusive). By locking, we put it into the distributed map with a time to live X and expiration time as value now() + X. So if it's not yet locked or already expired and evicted, we put it (back) into the map and return true, if it's still there, we refresh its expiration and return false. The return value will be used whether to add it to the queue or not, so we avoid having the same linking ids multiple times in any of the 2 queues.
I hope this clears it somewhat up, feel free to make any suggestions!

dbt commented on src/main/kotlin/com/openlattice/indexing/BackgroundLinkingIndexingService.kt:
```diff
@@ -284,23 +325,58 @@ class BackgroundLinkingIndexingService(
+
+    private fun refreshExpiration(linkingId: UUID) {
+        try {
+            linkingIndexingLocks.lock(linkingId)
```
why does this use hazel locks when the other lock methods do not?

you-leee replied:
Changed it to lock the whole `lockOrRefresh` process to avoid situations, where we start with expiration != null, but another process deletes it from the map and then the first process calls refresh and puts it back in

Commit:
LATTICE-2318 early return, fix condition

Commit:
LATTICE-2318 more human readable docs

Commit:
LATTICE-2318 lock whole getOrRefresh process

Commit:
LATTICE-2318 merge tasks, code reuse

Commit:
compile

Commit:
Merge branch 'develop' into bugfix/LATTICE-2318-linking-indexing-distributed-queue-destroyed

* develop:
  Woot
  compile
  Use streams for batching, mapping and summing instead of doing it manually
  Get rid of expiry job in favor of IMap TTLs Get rid of index on indexingLocks value elements

Commit:
Merge branch 'develop' into bugfix/LATTICE-2318-linking-indexing-distributed-queue-destroyed

* develop:
  use constants for args!
  use actual args :facepalm:
  fix logging for great perfz restructure conductorBootstrap
  checkstyle
  remove auth0 test
  fix test
  update Bootstrap
  Suppress FB Warning with reference to ticket for fixing this issue
  variable build version
  plugin update
  fix findbugs
  fix build maybe?

you-leee commented:
Update on the queries:
```
// query for un-indexing
SELECT linking_id,
       array_agg(ARRAY[entity_set_id, id]) AS entity_data_key,
       max(last_write) AS last_write
FROM ids
WHERE linking_id NOT IN
    (SELECT linking_id
     FROM ids
     WHERE linking_id IS NOT NULL
       AND VERSION > 0)
  AND linking_id IS NOT NULL
  AND last_index >= last_write
  AND last_link_index < last_write
GROUP BY linking_id LIMIT 3000

// query for indexing
WITH valid_linking_entities AS
  (SELECT linking_id,
          id,
          entity_set_id,
          last_write,
          last_index,
          last_link,
          last_link_index
   FROM ids
   WHERE linking_id IS NOT NULL
     AND VERSION > 0)
SELECT linking_id, array_agg(ARRAY[entity_set_id, id]) AS entity_data_key, max(last_write) AS last_write 
FROM valid_linking_entities 
WHERE last_index >= last_write 
	AND last_link >= last_write 
	AND last_link_index < last_write  
GROUP BY linking_id 
UNION ALL 
SELECT linking_id, array_agg(ARRAY[entity_set_id, id]) AS entity_data_key, max(last_write) AS last_write
FROM ids
WHERE linking_id IS NOT NULL
  AND VERSION <= 0
  AND last_index >= last_write
  AND last_link_index < last_write
  AND linking_id IN
    (SELECT linking_id
     FROM valid_linking_entities)
  AND entity_set_id IN
    (SELECT entity_set_id
     FROM valid_linking_entities)
GROUP BY linking_id LIMIT 3000
```

pckeyes commented on src/main/kotlin/com/openlattice/indexing/BackgroundLinkingIndexingService.kt:
```diff
@@ -103,54 +113,92 @@ class BackgroundLinkingIndexingService(
+                        .chunked(LINKING_INDEX_SIZE)
+                        .forEach { candidateBatch ->
+
+                            limiter.acquire()
```
as a novice at threads sorry if I'm wrong, but shouldn't this happen before/right after the submit call?

you-leee replied:
In this case, we use the limiter `Semaphore` to limit the number of jobs submitted to a certain number and not for locking the whole process

geekbeast commented on src/main/kotlin/com/openlattice/indexing/BackgroundLinkingIndexingService.kt:
```diff
@@ -103,54 +113,92 @@ class BackgroundLinkingIndexingService(
+    }
+
+    @Suppress("UNUSED")
+    @Scheduled(fixedRate = LINKING_INDEX_RATE)
```
Make this run forever in the background.

geekbeast commented on src/main/kotlin/com/openlattice/indexing/BackgroundLinkingIndexingService.kt:
```diff
@@ -103,54 +113,92 @@ class BackgroundLinkingIndexingService(
+                logger.info("Starting background linking $taskName task.")
+                val esw = Stopwatch.createStarted()
+
+                generateSequence(candidates::poll)
```
Should be take.

geekbeast reviewed (changes requested):
Use take a single non-scheduled thread.

Commit:
LATTICE-2318 fix feedback

Commit:
LATTICE-2318 more infinity loops, less logging

Commit:
LATTICE-2318 use poll with waiting to get elements, when count < LINKING_INDEX_SIZE

Commit:
Merge branch 'develop' into bugfix/LATTICE-2318-linking-indexing-distributed-queue-destroyed

* develop:
  remove redundant generics
  Make HazelcastMap generic-safe.

Commit:
LATTICE-2318 use ChunkedQueueSequence in background linking indexing, fix some warnings

Commit:
unused import

Commit:
LATTICE-2318 remove double logging and extra while loop

Commit:
Merge branch 'develop' into bugfix/LATTICE-2318-linking-indexing-distributed-queue-destroyed

* develop:
  LATTICE-2364 re-add Auth0Pod
  LATTICE-2364 create separate pod for hazelcast queues, remove duplicate pods

Commit:
Merge branch 'develop' into bugfix/LATTICE-2318-linking-indexing-distributed-queue-destroyed

* develop:
  use typesafe queue
  Typesafe queues

# Conflicts:
#	src/main/kotlin/com/openlattice/indexing/BackgroundLinkingIndexingService.kt
